From 36e89af697141a1094cf281285350b88b9db6b34 Mon Sep 17 00:00:00 2001
Message-ID: <36e89af697141a1094cf281285350b88b9db6b34.1767884020.git.sam@gentoo.org>
In-Reply-To: <71148ee5f0390fb3a35743a6fc45285b965384af.1767884019.git.sam@gentoo.org>
References: <71148ee5f0390fb3a35743a6fc45285b965384af.1767884019.git.sam@gentoo.org>
From: Indu Bhagat <indu.bhagat@oracle.com>
Date: Mon, 22 Dec 2025 00:54:44 -0800
Subject: [PATCH 18/41] [SFrame-V3] gas: sframe: use three states for location
 tracking

Up until now, for SFrame stack trace data generation (for default FDE
type), tracking of two states sufficed to distinguish between the two
cases:
  - the tracked entity is saved on a location on stack (tracked by
    SFRAME_FRE_ELEM_LOC_STACK)
  - the tracked entity is in register (tracked by
    SFRAME_FRE_ELEM_LOC_REG).

(In some sense, the distinction between the above two was the same as
saying SFRAME_FRE_ELEM_LOC_TRACKED or SFRAME_FRE_ELEM_LOC_UNTRACKED
respectively.)

Soon though, we will start to generate a new FDE type
SFRAME_FDE_TYPE_FLEX, where:
  - the tracked entity may be saved in a temporary register
  - the tracked entity may be saved at a "non-standard" location, e.g.,
    not a simple CFA+offset based location
  - and other cases

To effectively distinguish between the various states (necessary to
track for flex FDEs), define three states to track the location of each
tracked entity:
  - SFRAME_FRE_ELEM_LOC_NONE: the entity is not tracked
  - SFRAME_FRE_ELEM_LOC_REG: the entitiy is in a location based off a
    register
  - SFRAME_FRE_ELEM_LOC_STACK: the entity is in a located based off the
    CFA

While at it, rather than asserting in sframe_xlate_do_offset (), reset
the fp_reg state to SFRAME_FRE_REG_INVALID.  This is in preparation for
upcoming flex FDE generation patches.

TBD:
  - Add Co-Authored-By, Suggested-By Jens
  - Ideally the interfaces sframe_fre_set_ra_track,
    sframe_fre_set_fp_track need an uplift, but we can address that when
    time permits.

gas/
	* gen-sframe.c (sframe_xlate_do_offset): Reset other state.
	(sframe_xlate_do_same_value): Reset to SFRAME_FRE_ELEM_LOC_NONE.
	* gen-sframe.h (SFRAME_FRE_ELEM_LOC_REG): New definition.
	(SFRAME_FRE_ELEM_LOC_STACK): Likewise.
	(SFRAME_FRE_ELEM_LOC_NONE): Likewise.

---
[New in V1.  Not represent in RFC.]

[Changes in V2]
  - Moved out some stubs resetting ra_reg, ra_deref_p, fp_deref_p
    erroneously brought into this commit due a previous rebase.  These
    fields are introduced only in a later commit, and hence this patch
    would introduce build failures, making this not bisectable [Indu].
[End of changes in V2]

Signed-off-by: Sam James <sam@gentoo.org>
---
 gas/gen-sframe.c |  6 +++---
 gas/gen-sframe.h | 10 ++++++++--
 2 files changed, 11 insertions(+), 5 deletions(-)

diff --git a/gas/gen-sframe.c b/gas/gen-sframe.c
index 46efdd91636..a21756c82b9 100644
--- a/gas/gen-sframe.c
+++ b/gas/gen-sframe.c
@@ -1253,8 +1253,8 @@ sframe_xlate_do_offset (struct sframe_xlate_ctx *xlate_ctx,
   /* Ignore SP reg, as it can be recovered from the CFA tracking info.  */
   if (cfi_insn->u.ri.reg == SFRAME_CFA_FP_REG)
     {
-      gas_assert (cur_fre->fp_reg == SFRAME_FRE_REG_INVALID);
       sframe_fre_set_fp_track (cur_fre, cfi_insn->u.ri.offset);
+      cur_fre->fp_reg = SFRAME_FRE_REG_INVALID;
       cur_fre->merge_candidate = false;
     }
   else if (sframe_ra_tracking_p ()
@@ -1878,14 +1878,14 @@ sframe_xlate_do_same_value (const struct sframe_xlate_ctx *xlate_ctx,
 
   if (sframe_ra_tracking_p () && cfi_insn->u.r == SFRAME_CFA_RA_REG)
     {
-      cur_fre->ra_loc = SFRAME_FRE_ELEM_LOC_REG;
+      cur_fre->ra_loc = SFRAME_FRE_ELEM_LOC_NONE;
       cur_fre->ra_offset = 0;
       cur_fre->ra_undefined_p = false;
       cur_fre->merge_candidate = false;
     }
   else if (cfi_insn->u.r == SFRAME_CFA_FP_REG)
     {
-      cur_fre->fp_loc = SFRAME_FRE_ELEM_LOC_REG;
+      cur_fre->fp_loc = SFRAME_FRE_ELEM_LOC_NONE;
       cur_fre->fp_offset = 0;
       cur_fre->merge_candidate = false;
     }
diff --git a/gas/gen-sframe.h b/gas/gen-sframe.h
index eff089f6a24..fe7f3961fc9 100644
--- a/gas/gen-sframe.h
+++ b/gas/gen-sframe.h
@@ -31,8 +31,14 @@
       as_bad (format, __VA_ARGS__);            \
   } while (0)
 
-#define SFRAME_FRE_ELEM_LOC_REG		0
-#define SFRAME_FRE_ELEM_LOC_STACK	1
+/* The entity is not tracked.  */
+#define SFRAME_FRE_ELEM_LOC_NONE	0
+/* The location of the tracked entity is based on a register.  May or may not
+   involve dereferencing.  */
+#define SFRAME_FRE_ELEM_LOC_REG		1
+/* The location of the tracked entity is based on the CFA.  In theory, may or
+   may not involve dereferencing.  */
+#define SFRAME_FRE_ELEM_LOC_STACK	2
 
 /* An invalid register number.  */
 #define SFRAME_FRE_REG_INVALID		((unsigned int)-1)
-- 
2.52.0

